virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,284 @@ func (svc *googleService) insertSSHKey() error {
+	if err != nil {
+		return err
+	}
+
+	gDisk := &googlecloud.Disk{
+		Name:        disk.Name,
+		SizeGb:      int64(disk.DiskSizeGb),
+		Type:        diskType.SelfLink,
+		Description: disk.Description,
+	}
+
+	op, err := svc.service.Disks.Insert(svc.vm.Project, svc.vm.Zone, gDisk).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+}
+
+// attachDisk attaches a disk to an instance
+func (svc *googleService) attachDisk(disk *Disk) error {
+
+	gDisk, err := svc.getDisk(disk.Name)
+	if err != nil {
+		return err
+	}
+
+	gAttachedDisk := &googlecloud.AttachedDisk{
+		Source:     gDisk.SelfLink,
+		DeviceName: disk.Name,
+		AutoDelete: disk.AutoDelete,
+	}
+
+	op, err := svc.service.Instances.AttachDisk(svc.vm.Project, svc.vm.Zone,
+		svc.vm.Name, gAttachedDisk).Do()
+
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+}
+
+// detachDisk detaches a disk from an instance
+func (svc *googleService) detachDisk(disk *Disk) error {
+
+	var deviceName string
+
+	gDisk, err := svc.getDisk(disk.Name)
+	if err != nil {
+		return err
+	}
+	instance, err := svc.getInstance()
+	if err != nil {
+		return err
+	}
+
+	for _, attachedDisk := range instance.Disks {
+		if attachedDisk.Source == gDisk.SelfLink {
+			deviceName = attachedDisk.DeviceName
+			break
+		}
+	}
+
+	op, err := svc.service.Instances.DetachDisk(svc.vm.Project, svc.vm.Zone,
+		svc.vm.Name, deviceName).Do()
+
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+
+}
+
+// getFirewall gets details of a firewall
+func (svc *googleService) getFirewall() (*googlecloud.Firewall, error) {
+	return svc.service.Firewalls.Get(svc.vm.Project, svc.vm.Firewall).Do()
+}
+
+// addFirewallPorts adds new ports to a firewall
+func (svc *googleService) addFirewallPorts() error {
+
+	currFirewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	newRules := currFirewall.Allowed
+	for _, endpoint := range svc.vm.Endpoints {
+		newRules = append(newRules, &googlecloud.FirewallAllowed{
+			IPProtocol: endpoint.Protocol,
+			Ports:      endpoint.Ports,
+		})
+	}
+
+	return svc.patchFirewall(newRules)
+}
+
+// patchFirewall patches the firewall with the given allowed ports (rules)
+func (svc *googleService) patchFirewall(allowed []*googlecloud.FirewallAllowed) error {
+	newFirewall := &googlecloud.Firewall{
+		Allowed: allowed,
+	}
+	op, err := svc.service.Firewalls.Patch(svc.vm.Project, svc.vm.Firewall,
+		newFirewall).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForGlobalOperationReady(op.Name)
+}
+
+// removeFirewallPorts removes ports from a given firewall
+func (svc *googleService) removeFirewallPorts() error {
+	firewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	// Initialize new rules with the current one
+	// Refer to this link for structure of firewall resource
+	// https://cloud.google.com/compute/docs/reference/latest/firewalls
+	newRules := firewall.Allowed
```

Comment:
Please move all code of this func, till this statement below, where we are traversing newrules. Just to keep it in flow.